### DISCUSSION TITLE
external app to grab focused app app menus

### DISCUSSION THREAD

**Discussion by godbout** [godbout/Wooshy.docs#69] [Ideas]
here's the current train of thought:

1. currently you can target the app menu (left of menu bar) with Wooshy, and click the top menus
2. it's usually better to use the macOS keyboard shortcut `shift command /` and search, which grabs any menu or submenu available
3. i want to keep Wooshy targeted and specific: no list of results, etc. just grabbing WHAT YOU SEE on the screen. Dock hidden? don't grab it. menu bar hidden? don't grab it. 
4. i'm personally still a heavy user of Sublime Text and i love their command palette. i map the same shortcut to other editors i use (currently only Xcode). it'd be cool to have this for the whole macOS. i know it's possible to be done.

so the idea, another app (free, may or may not be open source). 

the only disadvantage i see from a user's perspective is (always the same): another global keyboard shortcut. but in this case Sublime Text also has one, so (in my case), the muscle memory is there. also `shift command p` would replace something related to printing (Page Setup?) that i personally never use.

any thoughts/advices?

**Comment by jannis-baum** [comment]
I'm completely with you on 3!

I don't use Sublime so maybe I'm missing something, but what would be the advantage of this command palette over using `shift command /`? What do you have mapped to the shortcut in Xcode?

Is there anything else that you would want in the macOS command palette aside from app menu items? If so, how would you decide what belongs into the command palette vs into something like Alfred? That said, if the case of "command palette vs native `shift command /`" is mostly about the command palette-like UI, wouldn't it be a good option to create an Alfred workflow to target app menus?

**Reply by godbout** [reply]
> I don't use Sublime so maybe I'm missing something, but what would be the advantage of this command palette over using `shift command /`?

good point. as far as i can tell, mostly none 😂️
maybe a better search? like fuzzy match etc.

> What do you have mapped to the shortcut in Xcode?

`Open Quicky...`
but actually i have `command p` there, not `shift command p`. not even sure why...

> Is there anything else that you would want in the macOS command palette aside from app menu items?

nope.

> If so, how would you decide what belongs into the command palette vs into something like Alfred? That said, if the case of "command palette vs native shift command /" is mostly about the command palette-like UI, wouldn't it be a good option to create an Alfred workflow to target app menus?

i think the Alfred issue for me is that i want to get this mapped to the Sublime Text `shift command p` (that has been copied by a lot of apps and sites around). i don't want Alfred to be mapped to that keyboard shortcut though. EDIT: that was a really dumb comment coz you can make the workflow to anything you want lol

but, well, you made me think i could map `shit command p` to the native macOS menu search. so that's it, problem solved 😂️ i'm gonna use this for a while and see, but that should do the trick.

**Reply by jannis-baum** [reply]
I've heard somewhere that with kV's shiny new API you could even make a simpler key open the `shift command /` in Normal mode!🙈

**Reply by godbout** [reply]
djeenius!

**Reply by godbout** [reply]
> but actually i have `command p` there, not `shift command p`. not even sure why...

ok so just for reference, it's coz `command p` is for `Go to Anything`, which is what i want in Xcode. not the whole command palette.

**Reply by jannis-baum** [reply]
Actually great to know that Xcode has `Open Quickly`, this makes me hate it a bit less😂

I couldn't find `Go to Anything`, I guess I'll get it with Ventura.

**Reply by godbout** [reply]
> Actually great to know that Xcode has `Open Quickly`, this makes me hate it a bit less😂

very useful.

> I couldn't find `Go to Anything`, I guess I'll get it with Ventura.

sorry, this is not in Xcode but in Sublime Text. what i wanted to mean is that `shift command p` in ST in the command palette, but `command p` is the `Go to Anything`, which is what i needed in Xcode. same functionality as `Open Quickly`, so i've remapped that one to `command p` instead of `shift command p`.

you tried AppCode (i think that's the name?) from JetBrains? personally i haven't, but there seems to be some other options than Xcode?

**Reply by jannis-baum** [reply]
> you tried AppCode (i think that's the name?) from JetBrains? personally i haven't, but there seems to be some other options than Xcode?

Wow, that's interesting. I thought completion for SwiftUI etc. was exclusive to Xcode... but anyways, my main issues with Xcode are that it's not Vim and that it's not in my terminal haha so I don't think AppCode could solve those. Interesting to know that other IDEs can get completion for Apple's app frameworks though. Thanks!

**Reply by godbout** [reply]
😅️

i may be completely wrong but isn't the Swift/SwiftUI language system (SIL? fuck i think i should get another coffee too 😂️) is open source or something. so other apps can use it and get the autocompletion etc.

you tried also Vim mode in Xcode? i was building/using kV before it came up, but i had a quick look when it came out. at the time lots of stuff weren't following the Vim way. heard people complain that it's been kinda dropped. no news since it came out. (i'm not gonna complain 😂️)

**Reply by jannis-baum** [reply]
Swift as a language is open source but as far as I know, SwiftUI, AppKit, UIKit, etc aren't. I can definitely get general Swift completion in Vim, but I don't think it exists for Apple's app frameworks.

About the "Vim mode" in Xcode: It's still a thing in Monterey, but using kV in Xcode is way superior😂 Xcode's Vim bindings are missing a ton of features.

**Reply by godbout** [reply]
> Swift as a language is open source but as far as I know, SwiftUI, AppKit, UIKit, etc aren't. I can definitely get general Swift completion in Vim, but I don't think it exists for Apple's app frameworks.

yeah, Swift is open source and SwiftUI is not. but i thought they had something open source related to the completion.

actually this is what i was talking about but maybe it doesn't include SwiftUI etc.: https://github.com/apple/sourcekit-lsp
maybe it's not even related at all to what we're talking about 😂️

> About the "Vim mode" in Xcode: It's still a thing in Monterey, but using kV in Xcode is way superior😂 Xcode's Vim bindings are missing a ton of features.

😅️
my bad. what i meant by "dropped" was the development. haven't checked but pretty sure it's there in Xcode 14. but no update.

**Reply by jannis-baum** [reply]
I tried `sourcekit-lsp` years ago and it didn't do the app frameworks back then but looks like now it actually works🙏 thanks for making me check again! No more Xcode for me now hehe

**Reply by godbout** [reply]
i guess you can stop your kV subscription now ☹️🥺️😂️

**Comment by aaronrolls** [comment]
Nova by Panic and VSCode also use Command Shift P for their command palette. You can also get a command palette type thing with an Alfred extension https://github.com/BenziAhamed/Menu-Bar-Search. You can map that extension to a shortcut.

<img width="600" alt="CleanShot2022-08-29at21 40 52@2x" src="https://user-images.githubusercontent.com/47949221/187172749-4f66be3e-9649-4523-96f5-f3e5446dde27.png">

Is that what you where meaning?

**Reply by godbout** [reply]
😨️🤯️

huh. yes. i didn't even check if there was an Alfred Workflow. i guess i just really wanted to add a little sister to kV and Wooshy 😂️

definitely many good options already. no need to build something. thank you all!

**Reply by aaronrolls** [reply]
No worries. Great idea, someone just beat you to it!

**Reply by godbout** [reply]
yeah, that would have been surprising if no one came up with that. i knew about Paletro but not the alternatives.

**Comment by haveanicedavid** [comment]
> i want to keep Wooshy targeted and specific: no list of results, etc. just grabbing WHAT YOU SEE on the screen. Dock hidden? don't grab it. menu bar hidden? don't grab it.

💯 

Personally, what i’d find most useful as an ‘add on’ feature to wooshy (or another app) is the abiltiy to see more of what’s on the screen / navigate around a bit. Often i’ll be typing in one area and want to jump to _something_ I know is withn another area, but isn’t on screen. Like if I’m typing in an email in mail and want to focus on the left preview bar thing so I can scroll through (using kV normal mode in an app like mail, becuase I can do useful things… select, delete etc), right now I don’t know a good way to get there. There probably is an app shortcut, but I don’t know it, and would have to learn all the shortcuts for all the apps to be able to accomplish that like I want.

Right now, when I active homerow’s scroll mode I can highlight that section and scroll, but can’t actually _go there_. I can exit homerow scroll and use wooshy to click on an individual email, but again that would click on an email and take me back to where I started. 

What i’d love is something like a more advanced version of the built-in macOS accessibility feature you get by htiting tab / shift-tab in accessibilty mode. I don’t like macos accessibilty mode and generally keep it off, but if there were some combination of that + homerow’s scroll mode, where I could hit a hotkey to highlight my curently focused area and navigate to other areas using vim movements and select the area without necessarily clicking into it, it’d be useful for me. 

That also might be super proprietary, and I probably didn’t explain it well. Maybe another way to phrase it is that i’d like a way to ‘click’ the area _ouside_ of clickable elements in order to select the window (or an an area of the app) itself so I can use the vim movements for navigating, when available. I also don’t expect you to build this, just to be clear 😄

**Reply by godbout** [reply]
> Right now, when I active homerow’s scroll mode I can highlight that section and scroll, but can’t actually _go there_. I can exit homerow scroll and use wooshy to click on an individual email, but again that would click on an email and take me back to where I started.

not sure i follow the whole thing, hence some questions after. but the first thing that comes in mind is: have you tried the Wooshy shortcut just to move the mouse without clicking? (`shift command enter`)

> What i’d love is something like a more advanced version of the built-in macOS accessibility feature you get by htiting tab / shift-tab in accessibilty mode. I don’t like macos accessibilty mode and generally keep it off, but if there were some combination of that + homerow’s scroll mode, where I could hit a hotkey to highlight my curently focused area and navigate to other areas using vim movements and select the area without necessarily clicking into it, it’d be useful for me.

is that VoiceOver or another thing? can you let me know how to enter this so that i can have a look? also maybe if you can share a screencast that'd help. thanks!

> Maybe another way to phrase it is that i’d like a way to ‘click’ the area ouside of clickable elements in order to select the window (or an an area of the app) itself so I can use the vim movements for navigating, when available.

again, not completely sure, but first thing that comes to mind is like when i tried to select `Scroll Areas` with Wooshy and it's not possible. so i'm not sure. i need to see that Accessibility feature you're talking above.

you make me think also about that thing in macOS where you can navigate buttons with the keyboard. some option you need to activate in the System Preferences. but kV should work there. but that's probably not what you mean. coz anyway it's much better to use Wooshy to target the UI Elements, so basically yeah i need to see what you mean with some video or something coz i don't get it sorry.
